src/bindings/rust/README.md

```diff
@@ -8,30 +8,151 @@
+
+    println!("Key with name {} has value {}", key.name(), key.value());
+
+    Ok(())
+}
+```
+
+Compared to the C-API, there are two distinct key types, `StringKey` and `BinaryKey`. With these, type mismatches such as calling `keyString` on a `BinaryKey` is not possible. The only difference between them is the type of value you can set and get from them. They are only wrappers over the `Key` from the C-API.
+
+Use a `BinaryKey` for setting arbitrary byte values.
+
+```rust
+extern crate elektra;
+use elektra::{BinaryKey, ReadableKey, WriteableKey};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let binary_content: [u8; 7] = [25, 34, 0, 254, 1, 0, 7];
+    let mut key = BinaryKey::new("user/test/rust")?;
+    key.set_value(&binary_content);
+    let read_content = key.value();
+
+    println!(
+        "Key with name {} holds bytes {:?}",
+        key.name(),
+        read_content
+    );
+
+    Ok(())
+}
+```
+
+The functionality of the keys is split into two traits, `ReadableKey` and `WritableKey`, which define methods that only read information from a key, and modify a key, respectively. For example, the method to retrieve metakeys only returns a key that implements `ReadableKey`, which is named `ReadOnly`. The keys returned cannot be modified in accordance to the design.
+
+### KeySet
+
+A KeySet is a set of StringKeys.
+
+- You can create an empty keyset with `new` or preallocate space for a number of keys with `with_capacity`.
+- It has two implementations of the `Iterator` trait, so you can iterate immutably or mutably.
+
+```rust
+extern crate elektra;
+use elektra::{KeyBuilder, KeySet, ReadableKey, StringKey};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Create a new KeySet with enough preallocated space for 5 keys
+    let mut keyset = KeySet::with_capacity(5);
+
+    // Append some keys
+    keyset.append_key(
+        KeyBuilder::<StringKey>::new("user/sw/app/#1/host")?
+            .value("localhost")
+            .build(),
+    )?;
+    keyset.append_key(
+        KeyBuilder::<StringKey>::new("user/sw/app/#1/port")?
+            .value("8080")
+            .build(),
+    )?;
+
+    // Iterate the keyset
+    keyset.rewind();
+    for key in keyset.iter() {
+        println!("Key ({}, {})", key.name(), key.value());
+    }
+
+    Ok(())
+    }
+```
+
+A `KeySet` only contains `StringKey`s, since they are far more prevalent than `BinaryKey`s. However since the underlying KeySet holds generic `Key`s, `BinaryKey`s can occur. You can cast between the two keys, by using the `From` trait. This is safe memory-wise, but can be unsafe if you cast a `BinaryKey` holding arbitrary bytes to a `StringKey`. You can use `is_string` or `is_binary` to find out whether the cast is safe.
+
+```rust
+let mut key = StringKey::new("user/test/language")?;
+
+// Cast the StringKey to BinaryKey
+let binary_key = BinaryKey::from(key);
+
+// And cast it back
+let string_key = StringKey::from(binary_key);
+```
```

Comment:
Did you already show all features of KeySets here?

Reply:
I linked to the KeySet example from the Readme, do you think that's enough?

Reply:
Putting a link sounds like a good plan.